lib/system_coordinator.rb

```diff
@@ -0,0 +1,174 @@
+require_relative 'room'
+require_relative 'date_range'
+require_relative 'reservation'
+require_relative 'block'
+require_relative 'no_availability_error'
+
+module Hotel
+  class SystemCoordinator    
+    attr_reader :rooms
+
+    @@next_block = 1
+
+    def initialize(room_quantity = 20)
+      @rooms = []   # Array.new(quantity){|i| Hotel::Room.new(i+1)}
+      build_rooms(room_quantity)
+    end
+
+    def build_rooms(room_quantity)
+      room_quantity.times do
+        @rooms << Hotel::Room.new(@rooms.length + 1)
+      end
+    end
+
+    def list_rooms
+      return rooms
+    end
+
+
+    def find_reservations_by_date(date)
+      reservations_by_date = []
+
+      rooms.each do |room|
+        room.bookings.each do |reservation|
+          reservations_by_date << reservation if reservation.date_range.include_date(date)
+        end
+      end
+
+      return reservations_by_date
+    end
+
+
+    def find_reservations_room_date(room_id, date_range)
+      selected_room = find_room(room_id)
+      reservations_room_date = selected_room.bookings.reject{|reservation|false == reservation.date_range.overlapping(date_range)}
+      return reservations_room_date    
+    end
+
+
+    def find_reservations_range(given_range)
+      reservations_range = []
+
+      rooms.each do |room|
+        room.bookings.each do |reservation|
+          reservations_range << reservation if reservation.date_range.overlapping(given_range)
+        end
+      end
+
+      return reservations_range
+    end
+
+
+    def find_available_rooms(given_range)
+      available_rooms = []
+      
+      rooms.each do |room|
+        if room.bookings.empty?
+          available_rooms << room
+        else
+          status = true
+          room.bookings.each do |item|
+            if item.block >= 0  #item.class == Hotel::Block 
+              status = false if item.date_range.overlapping(given_range) && false == item.date_range.exactly_matching(given_range) #it is not an exact match #method
+            elsif item.block < 0  #item.class == Hotel::Reservation
+              status = false if item.date_range.overlapping(given_range)
+            end
+          end
+          available_rooms << room if true == status
```

Comment:
You don't need to compare to `true` here since `status` is a boolean value:

```suggestion
          available_rooms << room if status
```